### DISCUSSION TITLE
Hash Table

### DISCUSSION THREAD

**Discussion by jangwooooo** [subak-ba/cs-study#34] [자료 구조]
## Hash Table
해시 테이블은 (Key, Value)로 데이터를 저장하는 자료구조다. 
각각의 키값에 해시함수를 적용해 배열의 고유한 index를 생성하고 저장한고, 이로 인해 빠른 검색속도를 제공한다.
여기서 실제 값이 저장되는 장소를 버킷 또는 슬롯이라고 한다.

예를 들어 (Key, Value)가 ("최장우", "수박")인 데이터를 크기가 13인 해시 테이블에 저장하면 index = hashFunction("최장우") % 13 연산을 통해 index 값을 계산한다. 그리고 array[index] = "수박" 으로 Value를 저장하게 된다.

<br/>

## 해시 충돌 
만약 "최장우"와 "오영기"를 해시 함수에 돌려서 나온 값이 동일하다면 문제가 생길 것이다.
해시 테이블에서는 충돌에 의한 문제를 분리 연결법(Separate Chaining)과 개방 주소법(Open Addressing) 크게 2가지로 해결하고 있다.

### 분리 연결법
![image](https://github.com/subak-ba/cs-study/assets/108380536/bdfc87a4-1967-4615-a6a3-84a01e596d96)
분리 연결법이란 동일한 버킷의 데이터에 대해 자료구조를 활용해 추가 메모리를 사용하여 다음 데이터의 주소를 저장하는 것이다. 
위의 그림과 같이 동일한 버킷으로 접근을 한다면 데이터들을 연결을 해서 관리해준다.
이러한 Chaining 방식은 해시 테이블의 확장이 필요없고 간단하게 구현이 가능하며, 손쉽게 삭제할 수 있다는 장점이 있다. 
하지만 데이터의 수가 많아지면 동일한 버킷에 Chaining되는 데이터가 많아지며 그에 따라 캐시의 효율성이 감소한다는 단점이 있다.

### 개방 주소법
개방 주소법이란 비어있는 해시 테이블의 공간을 활용하는 방법이다. 
Open Addressing을 구현하기 위한 대표적인 방법으로는 3가지 방식이 존재한다.

1. Linear Probing: 현재의 버킷 index로부터 고정폭 만큼씩 이동하여 차례대로 검색해 비어 있는 버킷에 데이터를 저장한다.
2. Quadratic Probing: 해시의 저장순서 폭을 제곱으로 저장하는 방식이다. 예를 들어 처음 충돌이 발생한 경우에는 1만큼 이동하고 그 다음 계속 충돌이 발생하면 2^2, 3^2 칸씩 옮기는 방식이다.
3. Double Hashing Probing: 해시된 값을 한번 더 해싱하여 해시의 규칙성을 없애버리는 방식이다. 해시된 값을 한번 더 해싱하여 새로운 주소를 할당하기 때문에 다른 방법들보다 많은 연산을 하게 된다.

![image](https://github.com/subak-ba/cs-study/assets/108380536/6f573377-f73b-4d8d-9007-4d6fdd4841e1)
Open Addressing에서 데이터를 삭제하면 삭제된 공간은 Dummy Space로 활용되는데, 그렇기 때문에 Hash Table을 재정리 해주는 작업이 필요하다.

<br/>

## 해시테이블 시간복잡도
각각의 Key값은 해시함수에 의해 고유한 index를 가지게 되어 바로 접근할 수 있으므로 평균 O(1)의 시간복잡도로 데이터를 조회할 수 있다. 
하지만 데이터의 충돌이 발생한 경우 Chaining에 연결된 리스트들까지 검색을 해야 하므로 O(N)까지 시간복잡도가 증가할 수 있다.
통계적으로 해시 테이블의 공간 사용률이 70% ~ 80%정도가 되면 해시의 충돌이 빈번하게 발생하여 성능이 저하되기 시작한다고 한다.

**Comment by onee-only** [comment]
분리 연결법 사진 자료가 배경이 어두워서 잘 안 보이는데 흰색 바탕으로 바꿔주실 수 있나요?

**Comment by onee-only** [comment]
개방 주소법의 여러 시나리오에 대해서 어떤 동작이 있는지 더 자세히 설명 해 주면 좋을 것 같아요
